entity-data-service-rx-client/src/main/java/org/hypertrace/entity/data/service/rxclient/EntityDataCachingClient.java

```diff
@@ -62,14 +68,17 @@
       Entity entity, UpsertCondition condition, Duration maximumUpsertDelay) {
     SingleSubject<Entity> singleSubject = SingleSubject.create();
     EntityKey entityKey = EntityKey.entityInCurrentContext(entity);
-    if (this.pendingEntityUpdates.containsKey(entityKey)) {
-      // Update the key - not strictly necessary (because the pending update holds the write data),
-      // but good hygiene
-      this.pendingEntityUpdates.put(entityKey, this.pendingEntityUpdates.get(entityKey));
+
+    // Acquire lock allowing multiple concurrent update adds, but only if no update executions
+    Lock lock = this.pendingUpdateStripedLock.get(entityKey).readLock();
+    try {
+      lock.lock();
+      this.pendingEntityUpdates
+          .computeIfAbsent(entityKey, unused -> new PendingEntityUpdate())
+          .addNewUpdate(entityKey, singleSubject, condition, maximumUpsertDelay);
```

Comment:
Trying to understand the code. Given the read lock, multiple threads will be allowed to go ahead and perform 2 operations concurrently:
1. Add a new entry within `pendingEntityUpdates` ConcurrentMap (if absent).
2. Mutate the `pendingEntityUpdate` instance fetched from the map as part of `addNewUpdate()`.

While step 1) is thread-safe given concurrentHashMap, step 2) is not thread-safe. Multiple threads may end up mutating `List<SingleObserver<Entity>> responseObservers` (since LinkedList is not thread-safe), along with overwriting other references like 
`private Disposable updateExecutionTimer;
    private Instant currentDeadline;
    private UpsertCondition condition;`

We need to synchronize the mutation of pendingEntityUpdate instance, right?

Reply:
Yeah, that's certainly fair. the issues we were seeing were due to interactions between 1 + 2, but as you point out multiple `addNewUpdate` calls are not safe currently. Will follow up with another change in there, hopefully to support concurrency without locking.

Reply:
Follow up in #114